runtime/compiler/p/codegen/PPCPrivateLinkage.cpp

```diff
@@ -1913,6 +1913,42 @@ int32_t TR::PPCPrivateLinkage::buildPrivateLinkageArgs(TR::Node
 	    addDependency(dependencies, NULL, (TR::RealRegister::RegNum)i, TR_VSX_SCALAR, cg());
       }
 
+
+
+   /*
+   
+   PPC ABI --
+   
+   AIX 32-bit BE:
+   AIX 64-bit BE:
+   Linux 64-bit BE: 
+   These are aix-style linkage. 
+   There is a TOC-base register (gr2) from function descriptor if that function is called. 
+   If the callee function is in libj9jitxx.so module, its TOC-base is recorded in J9VMThread jitTOC field already. 
+   All helper functions are in libj9jitxx.so module. So, you can load it from jitTOC field of J9VMThread (pointed to by metaDataRegister)
+  
+   Linux 32bit BE: Doesn't contain gr2. The register is reserved.
+
+   Linux 64-bit LE: there is TOC (gr2), but there is no function descriptor. 
+   TOC-base set-up is defined by ABI to go through a function's global-entry (relocation). 
+   The global-entry address should be set up in gr12. (although TOC-base is recorded in jitTOC field of J9VMThread as well ... none-C helpers are using it).
+
+   */
+   
+   if(linkage == TR_CHelper)
+   {
+      if(TR::Compiler->target.isLinux() && TR::Compiler->target.is64Bit() && TR::Compiler->target.cpu.isLittleEndian())
+      {
+         uintptrj_t targetAddress = (uintptrj_t)callSymbol->getMethodAddress();
+         loadConstant(cg(), callNode, (int64_t)targetAddress,dependencies->searchPreConditionRegister(TR::RealRegister::gr12));
```

Comment:
This needs a few instructions to materialize the address constant into gr12.
There is a better way:  the helper global-entry is in pseudo-TOC.  pseudo-TOC is based off gr16 (cg->getTOCBaseRegister()).  the index (offset) is 8 * helperSymbolIndex (I think).

Reply:
loadConstant seems to just generate one load instruction directly from gr16 into gr12.

Reply:
It might be, but that consumes pseudo-TOC space unnecessarily (i.e.  loadConstant doesn't know targetAddress being a helper address, treats it as a new constant, and allocates a new pseudo-TOC entry for it). On the other hand, targetAddress is already in pseudo-TOC for creating helper trampoline purpose.

Read code in runtime/compiler/runtime/Trampoline.cpp (createHelperTrampolines) and runtime/compiler/p/codegen/J9PPCInstruction.cpp (TR::PPCDepImmSymInstruction::generateBinaryEncoding()).

Reply:
you can verify the theory above by looking at the offset for the load you mentioned.  the offset is likely fairly big (more than 1200 for example).  the existing helper offset is likely less than 1200 (e.g.  helper index is less than 150).

Reply:
@zl-wang That's right. It did in fact create a new entry. I've made the respective changes, and tested them.